slsim/lens_pop.py

```diff
@@ -178,41 +187,57 @@ def draw_population(self, kwargs_lens_cuts, speed_factor=1):
                     source = self._sources.draw_source()
-                    gg_lens = Lens(
-                        deflector_dict=lens,
-                        source_dict=source,
-                        deflector_type=self._lens_galaxies.deflector_profile,
-                        variability_model=self._sources.variability_model,
-                        kwargs_variability=self._sources.kwargs_variability,
-                        sn_type=self.sn_type,
-                        sn_absolute_mag_band=self.sn_absolute_mag_band,
-                        sn_absolute_zpsys=self.sn_absolute_zpsys,
+                    _source = Source(
+                    source_dict=source,
+                    variability_model=self._sources.variability_model,
+                    kwargs_variability=self._sources.kwargs_variability,
+                    sn_type=self.sn_type,
+                    sn_absolute_mag_band=self.sn_absolute_mag_band,
+                    sn_absolute_zpsys=self.sn_absolute_zpsys,
+                    cosmo=self.cosmo,
+                    lightcurve_time=self.lightcurve_time,
+                    sn_modeldir=self.sn_modeldir,
+                    agn_driving_variability_model=self._sources.agn_driving_variability_model,
+                    agn_driving_kwargs_variability=self._sources.agn_driving_kwargs_variability,
+                    source_type=self._sources.source_type,
+                    light_profile=self._sources.light_profile,
+                    )
+                    lens_class = Lens(
+                        deflector_class=_lens,
+                        source_class=_source,
                         cosmo=self.cosmo,
                         test_area=test_area,
-                        source_type=self._sources.source_type,
                         los_config=self.los_config,
-                        light_profile=self._sources.light_profile,
-                        lightcurve_time=self.lightcurve_time,
-                        sn_modeldir=self.sn_modeldir,
-                        agn_driving_variability_model=self._sources.agn_driving_variability_model,
-                        agn_driving_kwargs_variability=self._sources.agn_driving_kwargs_variability,
                     )
                     # Check the validity of the lens system
-                    if gg_lens.validity_test(**kwargs_lens_cuts):
-                        gg_lens_population.append(gg_lens)
-                        # if a lens system passes the validity test, code should exit
-                        # the loop. so, n should be greater or equal to
-                        # num_sources_tested which will break the while loop
-                        # (instead of this one can simply use break).
-                        n = num_sources_tested
+                    if lens_class.validity_test(**kwargs_lens_cuts):
+                        valid_sources.append(_source)
+                    n += 1
+                if valid_sources:
```

Comment:
this is dangerous. Perhaps make the conditions explicit, i.e. if len(valid_sources) > 0: